### DISCUSSION TITLE
Let the rapiddiskd daemon to answer API requests without invoking rapiddisk executable

### DISCUSSION THREAD

**Discussion by matteotenca** [pkoutoupis/rapiddisk#125] [Ideas]
Hi there,

as in the title, I believe it would be nice to avoid the invocation of the `rapiddisk` executable from the daemon. Can this be achieved by moving some code to an external library? I am not much into this.

Just an idea!

Regards

**Comment by pkoutoupis** [comment]
Hello. That was the original thought until I discovered a memory leak in the libjansson code preventing it from properly freeing json objects. The cleanest way to do that was to fork into the executable and then exit with return code/stdout output. Reviewing my notes from 2 years ago...

I am very suspicious of this json_decref if statement (in jansson.h):
```
static JSON_INLINE void json_decref(json_t *json) {
    if (json && json->refcount != (size_t)-1 && JSON_INTERNAL_DECREF(json) == 0)
        json_delete(json);
}
```
I observed this memory leak in version 2.13.1 and older builds. Unfortunately, I never got around to proving the root cause of the problem.

**Reply by matteotenca** [reply]
I was checking out with `valgrind` and I found some memory leaks, not related to the json code, so I created a pull request for you to review (#129).
A part from that, as you said, there are memory leaks in the json related code. I gave a look at Jansson's docs and, well, it is a bit messy to release all the "references" created from many calls. But I believe that with a lot of patience it is possible to fix them all.

**Reply by pkoutoupis** [reply]
I see the PR. I will take a look shortly. As always, thank you! 😄

**Comment by pkoutoupis** [comment]
This was addressed in https://github.com/pkoutoupis/rapiddisk/pull/139.

**Comment by matteotenca** [comment]
Implemented in #139.

**Comment by pkoutoupis** [comment]
@matteotenca Thank you for closing this. And again, thank you for the wonderful and hard work to make this happen.
